doh-server/server.go

```diff
@@ -107,12 +111,46 @@ func (s *Server) Start() error {
 	if s.conf.Verbose {
 		servemux = handlers.CombinedLoggingHandler(os.Stdout, servemux)
 	}
+
+	var clientCAPool *x509.CertPool
+	if s.conf.TLSClientAuth && s.conf.CertCA != "" {
+		clientCA, err := ioutil.ReadFile(s.conf.CertCA)
+		if err != nil {
+			log.Fatalf("reading cert failed : %v", err)
+		}
+		clientCAPool = x509.NewCertPool()
+		clientCAPool.AppendCertsFromPEM(clientCA)
+		log.Println("Client cert loaded for TLS authentication")
+	}
+
 	results := make(chan error, len(s.conf.Listen))
 	for _, addr := range s.conf.Listen {
 		go func(addr string) {
 			var err error
+			log.Println("start server")
+
 			if s.conf.Cert != "" || s.conf.Key != "" {
-				err = http.ListenAndServeTLS(addr, s.conf.Cert, s.conf.Key, servemux)
+				if s.conf.TLSClientAuth && s.conf.CertCA != "" {
+					srvtls := &http.Server{
+						Handler: servemux,
+						Addr:    addr,
+						TLSConfig: &tls.Config{
+							ClientCAs:  clientCAPool,
+							ClientAuth: tls.RequireAndVerifyClientCert,
+							GetCertificate: func(info *tls.ClientHelloInfo) (certificate *tls.Certificate, e error) {
+								c, err := tls.LoadX509KeyPair(s.conf.Cert, s.conf.Key)
+								if err != nil {
+									fmt.Printf("Error loading key pair: %v\n", err)
+									return nil, err
+								}
+								return &c, nil
+							},
+						},
+					}
+					err = srvtls.ListenAndServeTLS("", "")
```

Comment:
Why you put `"", ""` here?

Reply:
I still want to know why `srvtls.ListenAndServeTLS("", "")` instead of `srvtls.ListenAndServeTLS(s.conf.Cert, s.conf.Key)`.

Maybe you are correct, but I don't know the reason. Would you please explain it to me?

Reply:
the ListenAndServeTLS in this situation can be called with "" because the certificate information has been set in the configuration part of the creation, here in the GetCertificate. The related code in the TLS lib is in server.go / ServeTLS. If a certificate is provided the incorporation is started again which is not efficient.

Reply:
Understand. Thanks.

Reply:
Thanks for the explanation.